llvm/lib/Target/AIE/AIEAddressSpaceLowering.cpp

```diff
@@ -0,0 +1,113 @@
+//===--------- AIEAddressSpaceLowering.cpp - Retype pointer to P0  --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-addrspace-lowering"
+
+using namespace llvm;
+
+static const char AIE_POINTER_RETYPE[] = "AIE Address Space Lowering Pass";
+
+namespace {
+
+class AIEAddressSpaceLowering : public MachineFunctionPass {
+public:
+  static char ID;
+  AIEAddressSpaceLowering() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE.
```

Comment:
I could do with some more comment. What is the function of CSE here? I can imagine cse-ing default address space expressions with specific ones?